heft-plugins/heft-esbuild-plugin/src/EsbuildPlugin.ts

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { ITerminal, Import } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  HeftSession,
+  ICompileSubstage,
+  IBuildStageContext,
+  IHeftPlugin,
+  ScopedLogger
+} from '@rushstack/heft';
+
+import { configurationFileLoader, IEsbuildConfigurationJson } from './EsbuildConfigLoader.js';
+import { BuildFailure, BuildResult } from 'esbuild';
+
+const esbuild: typeof import('esbuild') = Import.lazy('esbuild', require);
+
+const PLUGIN_NAME: string = 'EsbuildPlugin';
+
+interface IEsbuildConfigurationFileCacheEntry {
+  configurationFile: IEsbuildConfigurationJson | undefined;
+}
+
+/**
+ * @internal
+ */
+export class EsbuildPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  private _esbuildConfigurationFileCache: Map<string, IEsbuildConfigurationFileCacheEntry> = new Map<
+    string,
+    IEsbuildConfigurationFileCacheEntry
+  >();
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    const logger: ScopedLogger = heftSession.requestScopedLogger('esbuild');
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(PLUGIN_NAME, async () => {
```

Comment:
Do we want this in compile or bundle?

Reply:
This is a fundamental question. ESBuild can be used both as a compiler (for stripping Typescript to produce JS) and/or as a bundler.

Right now our primary use-case has been 'both', i.e. we compile the TS to JS with ESbuild and also bundle the output in one step. I attached Esbuild to the compile task in this case to allow us to do that compilation before running the Workbox plugin which is attached to the afterCompile hook. We basically then do not have a bundle task essentially since the output of our compile step is effectively bundled already.

I understand this is probably a little bit of a deviation from the traditional flow though so am open to other ideas of how to orchestrate this.